src/components/Dropdown/Dropdown.tsx

```diff
@@ -143,107 +148,144 @@ const Dropdown = React.forwardRef<HTMLDivElement, DropdownProps>(
 
-    useEffect(() => {
-      const handleKeyDown = (event: KeyboardEvent) => {
-        if (event.key === 'Escape') {
-          setIsOpen(false);
+    const handleEscape = useCallback(() => {
+      const escapeKeyDown = (event: KeyboardEvent) => {
+        if (event.code === 'Escape') {
+          switchDropDownOpenState();
         }
       };
 
       const dropdownElement = dropdownRef.current;
       if (dropdownElement) {
-        dropdownElement.addEventListener('keydown', handleKeyDown);
+        dropdownElement.addEventListener('keydown', escapeKeyDown);
+
+        return () => {
+          dropdownElement.removeEventListener('keydown', escapeKeyDown);
+        };
+      }
+    }, [dropdownRef, switchDropDownOpenState]);
+
+    const handleClickOpenClose = useCallback(() => {
+      if (disabled) {
+        return;
       }
 
-      return () => {
-        if (dropdownElement) {
-          dropdownElement.removeEventListener('keydown', handleKeyDown);
+      setIsOpen((prev) => {
+        const newState = !prev;
+        if (newState) {
+          handleEscape();
         }
-      };
-    }, [dropdownRef]);
+        return newState;
+      });
+      setClickEnabled(false);
+    }, [disabled, handleEscape]);
 
-    useEffect(() => {
-      const currentIndex = selectableOptions.findIndex(
-        (op) => op.id === selectedValue?.id,
-      );
-      if (currentIndex !== -1) {
-        const selectedElement = document.querySelector(
-          `[data-index="${currentIndex}"]`,
-        );
-        if (selectedElement) {
-          (selectedElement as HTMLDivElement).focus();
+    const handleKeyDownOpenClose = useCallback(
+      (event: React.KeyboardEvent<HTMLDivElement>) => {
+        if (event.code === 'Space' || event.code === 'Enter') {
+          event.preventDefault();
+          setClickEnabled(false);
+          handleClickOpenClose();
         }
+      },
+      [handleClickOpenClose],
+    );
+
+    const handleKeyDownUpOptions = useCallback(
+      (
+        event: React.KeyboardEvent<HTMLDivElement>,
+        option: OptionsType,
+        index: number,
+      ) => {
+        if (event.code === 'Space' || event.code === 'Enter') {
+          event.preventDefault();
+          handleChange(option);
+        }
+        if (event.code === 'ArrowDown') {
+          event.preventDefault();
+          const nextElement = document.querySelector(
+            `[data-index="${index + 1}"]`,
+          );
+          if (nextElement) {
+            (nextElement as HTMLDivElement).focus();
+          }
+        }
+        if (event.code === 'ArrowUp') {
+          event.preventDefault();
+          const previousElement = document.querySelector(
+            `[data-index="${index - 1}"]`,
+          );
+          if (previousElement) {
+            (previousElement as HTMLDivElement).focus();
+          }
+        }
+      },
+      [handleChange],
+    );
+
+    useEffect(() => {
```

Comment:
Is this useEffect really needed? Seems like it could be converted to a callback instead

Reply:
useEffect is gone.
However, my solution for this is to use a Timeout() in the checkCurrentIndex function. This because so I can get the div elements when the page has been rendered.

Other change is useRef for option elements and a function to handle selectableRefs.